File: packages/portal-domain/src/local-ai-runtime-panel.ts
import type { PortalActivityMemoryGraphReadModel } from './activity-memory-graph';
import {
  PortalAgentEvent as AgentEvent,
  PortalAgentProtocolField,
  type PortalRouteEventRecord,
} from './portal-contract-adapter';
import { decodePortalDetailValue, type PortalDetailValue } from './portal-contract-text-contracts';
import { decodeDisplayText, type DisplayText } from './display-text';
import { PortalDetails } from './details';
import {
  localAiRuntimePanelTemplate,
  type LocalAiHouseholdJobInput,
  type LocalAiMemoryGraphInput,
  type LocalAiRemoteAssistantBoundaryInput,
  type LocalAiRuntimePanelCardKind,
  type LocalAiRuntimePanelFieldKey,
  type LocalAiRuntimeStatusInput,
} from './local-ai-runtime-panel.generated';

export type LocalAiRuntimePanelDetail = {
  readonly label: DisplayText;
  readonly value: PortalDetailValue;
};

export type LocalAiRuntimePanelCard = {
  readonly title: DisplayText;
  readonly details: readonly LocalAiRuntimePanelDetail[];
};

export type LocalAiRuntimePanelIntent = {
  readonly eyebrow: DisplayText;
  readonly title: DisplayText;
  readonly body: DisplayText;
  readonly emptyMessage: DisplayText;
  readonly emptyStatus: PortalDetailValue;
  readonly productClaim: PortalDetailValue;
  readonly summaryDetails: readonly LocalAiRuntimePanelDetail[];
  readonly cards: readonly LocalAiRuntimePanelCard[];
};

const LocalAiRuntimePanelCardTitles: Readonly<Record<LocalAiRuntimePanelCardKind, DisplayText>> = {
  'runtime-status': decodeDisplayText('Local AI runtime status'),
  'household-job': decodeDisplayText('Household AI job activity'),
  'memory-graph': decodeDisplayText('Cited memory and graph evidence'),
  'remote-assistant-boundary': decodeDisplayText('Remote assistant boundary'),
};

const LocalAiRuntimePanelFieldLabels: Readonly<Record<LocalAiRuntimePanelFieldKey, DisplayText>> = {
  eventId: PortalDetails.EventId,
  sentAt: PortalDetails.SentAt,
  runtimeReference: PortalDetails.RuntimeReference,
  provider: PortalDetails.Provider,
  model: PortalDetails.Model,
  loadState: PortalDetails.LoadState,
  capability: PortalDetails.Capability,
  resourceClass: PortalDetails.ResourceClass,
  degradedState: PortalDetails.DegradedState,
  privacyMode: PortalDetails.PrivacyMode,
  executionState: PortalDetails.ExecutionState,
  reason: PortalDetails.Reason,
  requestId: PortalDetails.RequestId,
  status: PortalDetails.Status,
  state: PortalDetails.State,
  providerSource: PortalDetails.ProviderSource,
  custody: PortalDetails.Custody,
  policyReadiness: PortalDetails.PolicyReadiness,
  adapterBoundary: PortalDetails.AdapterBoundary,
  lastChecked: PortalDetails.LastChecked,
  lastObserved: PortalDetails.LastObserved,
  decisionSource: PortalDetails.DecisionSource,
  productClaim: PortalDetails.ProductClaim,
  generatedAt: PortalDetails.GeneratedAt,
  graphNodes: PortalDetails.GraphNodes,
  graphEdges: PortalDetails.GraphEdges,
  graphOmittedEdges: PortalDetails.GraphOmittedEdges,
  evidenceReferences: PortalDetails.EvidenceReferences,
  deletedEvidence: PortalDetails.DeletedEvidence,
  rowCount: PortalDetails.RowCount,
};

export function createLocalAiRuntimePanelIntent(
  runtimeEvent: PortalRouteEventRecord | null,
  lanAiJobEvent: PortalRouteEventRecord | null,
  memoryGraphReadModel: PortalActivityMemoryGraphReadModel | null = null,
  parentAssistantBoundaryEvent: PortalRouteEventRecord | null = null
): LocalAiRuntimePanelIntent {
  const template = localAiRuntimePanelTemplate({
    runtimeStatus: normalizeRuntimeStatus(runtimeEvent),
    householdJob: normalizeHouseholdJob(lanAiJobEvent),
    memoryGraph: normalizeMemoryGraph(memoryGraphReadModel),
    remoteAssistantBoundary: normalizeRemoteAssistantBoundary(parentAssistantBoundaryEvent),
  });

  return {
    eyebrow: decodeDisplayText('Local child-device AI'),
    title: decodeDisplayText('AI jobs and runtime activity'),
    body: decodeDisplayText('Service-reported local AI status and LAN job rows only.'),
    emptyMessage: decodeDisplayText('No local AI runtime or job event has been reported yet.'),
    emptyStatus: decodePortalDetailValue('not-reported'),
    productClaim: decodePortalDetailValue('runtime-read-model-only'),
    summaryDetails: [
      {
        label: PortalDetails.Status,
        value: decodePortalDetailValue(template.summaryStatus),
      },
      {
        label: PortalDetails.ReadModelRows,
        value: decodePortalDetailValue(template.summaryReadModelRows),
      },
      {
        label: PortalDetails.ProductClaim,
        value: decodePortalDetailValue(template.summaryProductClaim),
      },
    ],
    cards: template.cards.map((card) => ({
      title: LocalAiRuntimePanelCardTitles[card.kind],
      details: card.details.map((detail) => ({
        label: LocalAiRuntimePanelFieldLabels[detail.fieldKey],
        value: decodePortalDetailValue(detail.value),
      })),
    })),
  };
}

function normalizeRuntimeStatus(event: PortalRouteEventRecord | null): LocalAiRuntimeStatusInput | null {
  if (event?.event !== AgentEvent.LocalAiRuntimeStatusReported) {
    return null;
  }
  return {
    eventId: stringOrNull(event.eventId),
    sentAt: stringOrNull(event.sentAt),
    runtimeReference: payloadValue(event, PortalAgentProtocolField.LocalAiRuntimeReferenceId),
    provider: payloadValue(event, PortalAgentProtocolField.LocalAiProviderId),
    model: payloadValue(event, PortalAgentProtocolField.LocalAiModelId),
    loadState: payloadValue(event, PortalAgentProtocolField.LoadState),
    capability: payloadValue(event, PortalAgentProtocolField.LocalAiCapabilityFlags),
    resourceClass: payloadValue(event, PortalAgentProtocolField.LocalAiResourceClass),
    degradedState: payloadValue(event, PortalAgentProtocolField.LocalAiDegradedState),
    privacyMode: payloadValue(event, PortalAgentProtocolField.LocalAiPrivacyMode),
    executionState: payloadValue(event, PortalAgentProtocolField.LocalAiExecutionState),
    reason: payloadValue(event, PortalAgentProtocolField.LocalAiUnavailableReason),
  };
}

function normalizeHouseholdJob(event: PortalRouteEventRecord | null): LocalAiHouseholdJobInput | null {
  if (event?.event !== AgentEvent.LanAiJobReported) {
    return null;
  }
  return {
    eventId: stringOrNull(event.eventId),
    sentAt: stringOrNull(event.sentAt),
    requestId: payloadValue(event, PortalAgentProtocolField.LanAiJobId),
    status: payloadValue(event, PortalAgentProtocolField.LanAiJobStatus),
    state: payloadValue(event, PortalAgentProtocolField.LanAiJobState),
    provider: payloadValue(event, PortalAgentProtocolField.LocalAiProviderId),
    providerSource: payloadValue(event, PortalAgentProtocolField.LocalAiProviderSource),
    capability: payloadValue(event, PortalAgentProtocolField.LocalAiCapabilityFlags),
    resourceClass: payloadValue(event, PortalAgentProtocolField.LocalAiResourceClass),
    loadState: payloadValue(event, PortalAgentProtocolField.LocalAiAdapterReadinessState),
    privacyMode: payloadValue(event, PortalAgentProtocolField.LocalAiPrivacyMode),
    custody: payloadValue(event, PortalAgentProtocolField.LanAiProviderCustodyLabel),
    policyReadiness: payloadValue(event, PortalAgentProtocolField.LanAiProviderRoutingState),
    adapterBoundary: payloadValue(event, PortalAgentProtocolField.ClaimBoundary),
    leaseId: payloadValue(event, PortalAgentProtocolField.LanControllerLeaseId),
    lastChecked: payloadValue(event, PortalAgentProtocolField.LanControllerLeaseIssuedAt),
    lastObserved: payloadValue(event, PortalAgentProtocolField.LanControllerLeaseExpiresAt),
    decisionSource: payloadValue(event, PortalAgentProtocolField.LanParentAuthority),
    executionState: payloadValue(event, PortalAgentProtocolField.LocalAiExecutionState),
    reason: payloadValue(event, PortalAgentProtocolField.Reason),
  };
}

function normalizeMemoryGraph(readModel: PortalActivityMemoryGraphReadModel | null): LocalAiMemoryGraphInput | null {
  if (readModel === null) {
    return null;
  }
  return {
    custody: stringOrNull(readModel.custody),
    capabilityStatus: stringOrNull(readModel.capabilityStatus),
    generatedAt: stringOrNull(readModel.generatedAt),
    returnedNodeCount: readModel.returnedNodeCount,
    returnedEdgeCount: readModel.returnedEdgeCount,
    omittedEdgeCount: readModel.omittedEdgeCount,
    degradedReasons: [...readModel.degradedReasons],
    evidenceReferenceIds: collectMemoryGraphEvidenceReferenceIds(readModel),
  };
}

function normalizeRemoteAssistantBoundary(
  event: PortalRouteEventRecord | null
): LocalAiRemoteAssistantBoundaryInput | null {
  if (
    event?.event !== AgentEvent.ParentAssistantAnswerReported &&
    event?.event !== AgentEvent.ParentAssistantProviderDegraded &&
    event?.event !== AgentEvent.ParentAssistantErrorReported
  ) {
    return null;
  }
  return {
    eventId: stringOrNull(event.eventId),
    sentAt: stringOrNull(event.sentAt),
    requestId: payloadValue(event, PortalAgentProtocolField.ParentAssistantRequestId),
    state: payloadValue(event, PortalAgentProtocolField.ParentAssistantAnswerState),
    provider: payloadValue(event, PortalAgentProtocolField.ParentAssistantProviderRoute),
    adapterBoundary: payloadValue(event, PortalAgentProtocolField.ParentAssistantApiProviderBoundary),
    policyReadiness: payloadValue(event, PortalAgentProtocolField.ParentAssistantApiAuthorizationState),
    custody: payloadValue(event, PortalAgentProtocolField.ParentAssistantApiCustodyLabel),
    deletedEvidence: payloadValue(event, PortalAgentProtocolField.ParentAssistantApiDeletionState),
    privacyMode: payloadValue(event, PortalAgentProtocolField.ParentAssistantApiRetentionState),
    evidenceReferences: payloadValue(event, PortalAgentProtocolField.ParentAssistantEvidenceSummary),
    rowCount: payloadValue(event, PortalAgentProtocolField.ParentAssistantCitationCount),
  };
}

function collectMemoryGraphEvidenceReferenceIds(readModel: PortalActivityMemoryGraphReadModel): string[] {
  const refs: string[] = [];
  for (const node of readModel.nodes) {
    for (const evidenceRef of node.trace.sourceEvidenceReferences) {
      refs.push(evidenceRef.evidenceReferenceId);
    }
  }
  for (const edge of readModel.edges) {
    for (const evidenceRef of edge.trace.sourceEvidenceReferences) {
      refs.push(evidenceRef.evidenceReferenceId);
    }
  }
  return refs;
}

function payloadValue(event: PortalRouteEventRecord, key: string): string | null {
  return scalarString(event.payload?.[key]);
}

function scalarString(value: unknown): string | null {
  if (typeof value === 'string') {
    return value;
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  return null;
}

function stringOrNull(value: unknown): string | null {
  return typeof value === 'string' ? value : null;
}

File: crates/parent-runtime-core/tests/unit/local_ai_runtime_panel.rs
use ocentra_parent_runtime_core::local_ai_runtime_panel::{
    local_ai_runtime_panel_typescript, project_local_ai_runtime_panel, LocalAiHouseholdJobInput,
    LocalAiMemoryGraphInput, LocalAiRemoteAssistantBoundaryInput, LocalAiRuntimePanelInput,
    LocalAiRuntimeStatusInput,
};

#[test]
fn local_ai_runtime_panel_projects_runtime_and_household_rows() {
    let input = LocalAiRuntimePanelInput {
        runtime_status: Some(LocalAiRuntimeStatusInput {
            event_id: Some("evt-local-ai-runtime"),
            sent_at: Some("2026-06-07T19:15:00Z"),
            runtime_reference: Some("runtime-child-device-1"),
            provider: Some("local-provider-1"),
            model: Some("screen-local-vlm-v1"),
            load_state: Some("loaded"),
            capability: Some("ocr,vision"),
            resource_class: Some("gpu"),
            degraded_state: Some("ready"),
            privacy_mode: Some("local-only"),
            execution_state: Some("ready"),
            reason: None,
        }),
        household_job: Some(LocalAiHouseholdJobInput {
            event_id: Some("evt-lan-ai-job"),
            sent_at: Some("2026-06-07T19:15:02Z"),
            request_id: Some("lan-ai-job-1"),
            status: Some("claimed"),
            state: Some("worker-running"),
            provider: Some("household-desktop-provider"),
            provider_source: Some("trusted-household-desktop"),
            capability: Some("screen-hard-visual-analysis"),
            resource_class: Some("gpu"),
            load_state: Some("ready"),
            privacy_mode: Some("local-lan-redacted"),
            custody: Some("local-lan-redacted"),
            policy_readiness: Some("authorized-result"),
            adapter_boundary: Some("claim-lease-child-owned-job"),
            lease_id: Some("lease-screen-ai-1"),
            last_checked: Some("2026-06-07T19:15:01Z"),
            last_observed: Some("2026-06-07T19:20:01Z"),
            decision_source: Some("child-agent-local-policy-authority"),
            execution_state: Some("running"),
            reason: None,
        }),
        ..LocalAiRuntimePanelInput::default()
    };

    let panel = project_local_ai_runtime_panel(&input);

    assert_eq!(panel.summary_status, "reported");
    assert_eq!(panel.summary_read_model_rows, "2");
    assert_eq!(
        panel.summary_product_claim,
        "no-model-quality-or-enforcement-claim"
    );
    assert_eq!(panel.cards.len(), 2);
    assert_eq!(panel.cards[0].kind, "runtime-status");
    assert_eq!(panel.cards[1].kind, "household-job");
    assert!(panel.cards[0]
        .details
        .iter()
        .any(|detail| detail.field_key == "model" && detail.value == "screen-local-vlm-v1"));
    assert!(panel.cards[1].details.iter().any(|detail| {
        detail.field_key == "providerSource" && detail.value == "trusted-household-desktop"
    }));
    assert!(panel.cards[1]
        .details
        .iter()
        .any(|detail| detail.field_key == "productClaim"
            && detail.value == "worker-only-child-agent-authority"));
}

#[test]
fn local_ai_runtime_panel_projects_memory_graph_and_remote_boundary() {
    let input = LocalAiRuntimePanelInput {
        memory_graph: Some(LocalAiMemoryGraphInput {
            custody: Some("child-device-query-store"),
            capability_status: Some("ready"),
            generated_at: Some("2026-06-07T19:16:00Z"),
            returned_node_count: 1,
            returned_edge_count: 1,
            omitted_edge_count: 0,
            degraded_reasons: Vec::new(),
            evidence_reference_ids: vec![
                "evidence-screen-summary-1",
                "evidence-screen-summary-1",
            ],
        }),
        remote_assistant_boundary: Some(LocalAiRemoteAssistantBoundaryInput {
            event_id: Some("evt-parent-assistant-boundary"),
            sent_at: Some("2026-06-07T19:17:00Z"),
            request_id: Some("remote-assistant-request-1"),
            state: Some("ready-answer"),
            provider: Some("remote-api-report-only"),
            adapter_boundary: Some("parent-authorized-report-bundle"),
            policy_readiness: Some("parent-authorized"),
            custody: Some("parent-owned-local-storage"),
            deleted_evidence: Some("raw-model-output-not-retained"),
            privacy_mode: Some("report-summary-only"),
            evidence_references: Some("evidence-screen-summary-1"),
            row_count: Some("1"),
        }),
        ..LocalAiRuntimePanelInput::default()
    };

    let panel = project_local_ai_runtime_panel(&input);

    assert_eq!(panel.cards.len(), 2);
    assert_eq!(panel.cards[0].kind, "memory-graph");
    assert_eq!(panel.cards[1].kind, "remote-assistant-boundary");
    assert!(panel.cards[0]
        .details
        .iter()
        .any(|detail| detail.field_key == "evidenceReferences"
            && detail.value == "evidence-screen-summary-1"));
    assert!(panel.cards[1]
        .details
        .iter()
        .any(|detail| detail.field_key == "adapterBoundary"
            && detail.value == "parent-authorized-report-bundle"));
    assert!(panel.cards[1]
        .details
        .iter()
        .any(|detail| detail.field_key == "productClaim"
            && detail.value == "remote-assistant-report-only-local-policy-authority"));
}

#[test]
fn local_ai_runtime_panel_keeps_missing_inputs_visible_without_success() {
    let panel = project_local_ai_runtime_panel(&LocalAiRuntimePanelInput::default());

    assert_eq!(panel.cards, Vec::new());
    assert_eq!(panel.summary_status, "not-reported");
    assert_eq!(panel.summary_read_model_rows, "0");
}

#[test]
fn local_ai_runtime_panel_generated_typescript_stays_checked_in() {
    let checked_in = include_str!(
        "../../../../packages/portal-domain/src/generated/local-ai-runtime-panel.ts"
    );

    assert_eq!(checked_in, local_ai_runtime_panel_typescript());
}
